---
title: "predictions"
author: "Miriam"
date: "8/24/2022, changed nov2022"
output: html_document
---
This script uses the glmnet package in order to predict putative enhacers which contribute highly to a target gene. For this a gene name is chosen and the window, which determines the putative enhancers used as inputs.


# bundle all information for one gene 
(1) plot the GEX of that gene
(2) plot the promoter(s) with the scores and if there are predictions on TF

```{r,echo=F,warning=F,message=FALSE}
#libraries
library(glmnet)
library(GenomicRanges)
library(rtracklayer)
library(scales)
library(dplyr)
library(ggplot2)
library(Biostrings)
library(vip) 

source("functions_regulation.R", local = knitr::knit_global())
```

load data
```{r,echo=F}
#load data
dir <- "output/archetypes_16dec2022_allgenes_doub_yolk_removed_rank20"

#DGE
genes_dir <- file.path(dir,"genes")
GENES = as.character(read.table(file.path(genes_dir,"features.txt"))$x)
DGE = read.table(file.path(genes_dir,"input_data.txt"))

#dca
DCA <-read.table(file.path(dir,"prom_enh/input_data.txt"))
  
#enhancers
peaks_dir <- "output/annotate_ATAC_peaks"
enhancers <- read.table(file.path(peaks_dir,"peaks_enhancers.txt"), header = T, stringsAsFactors = F)
eDCA <- DCA[enhancers$name,]
g_eCRM <-makeGRangesFromDataFrame(enhancers ,keep.extra.columns = T)

#promoters
promoters <- read.table(file.path(peaks_dir,"peaks_promoters.txt"), header = T, stringsAsFactors = F)
pDCA <- DCA[promoters$name,]
g_pCRM <-makeGRangesFromDataFrame(promoters ,keep.extra.columns = T)

#annotations dm6
gtf <- import.gff("input/Drosophila_melanogaster.BDGP6.28.100.chr.gtf")
seqlevelsStyle(gtf) <- "UCSC"

#loacations for the cells
locations = read.table('input/geometry.txt',nrows = 3039, skip = 1,col.names = c("x","y","z"))

#TF_list
tf_dir <- "output/virtual_tfbs"
TF_list <- read.table(file.path(tf_dir, "TF_list.txt"),stringsAsFactors = F)
tfbs_hits <- read.table(file.path(tf_dir, "tfbs_hits_scaled.txt"))
#rownames(tfbs_hits)<- read.table(file.path("output/archetypes_28october2022_allgenes/promoter_enhancer/features.txt"),stringsAsFactors = F)$x
rownames(tfbs_hits)<- read.table(file.path("output/archetypes_16dec2022_allgenes_doub_yolk_removed_rank20/prom_enh/features.txt"),stringsAsFactors = F)$x

e_tfbs_hits <- tfbs_hits[g_eCRM$name,]
p_tfbs_hits <- tfbs_hits[g_pCRM$name,]
#rownames(p_tfbs_hits) <- pCRM
```

specify window size and gene
```{r,echo=F}
distance <- 80000 #distance inspected up and downstream(so total doubled)

# parameters and values 
testgene<- "FBgn0000606" #eve
testgene <- "FBgn0001325" # Kr
testgene <- "FBgn0003145" #prd
testgene <- "FBgn0000577" # engrailed
testgene <- "FBgn0001325" #kr
testgene <- "FBgn0011656" #mef2
testgene <- "FBgn0261930" #vnd
testgene<- "FBgn0010323" #gsm
testgene <- "FBgn0020912"#ptx1
testgene <- "FBgn0020307" #dve
testgene<-"FBgn0003900" #twst
testgene <- "FBgn0003274" #ihgh expressed
testgene<- "FBgn0004053" #zen
testgene <-"FBgn0003900" #twist
testgene <- "FBgn0003448" #snail
testgene <-"FBgn0003463" #sog2
testgene<- "FBgn0001320" #kni
testgene <- "FBgn0003944" #ubx
testgene <- "FBgn0011236" #ken
testgene <- "FBgn0001148" #gsb
testgene <- "FBgn0004919" #gol
testgene<- "FBgn0010109" #dpn
testgene <- "FBgn0024250" #brk
testgene <- "FBgn0003463" #sog
testgene <- "FBgn0000251" #caudal
testgene <- "FBgn0025776" #ind
testgene <- "FBgn0051721" #trim9
testgene <- "FBgn0011746" #ana

```

#fit the model
```{r,echo=F}
#get the five_primes for setting the locations
five_prime <- gtf[gtf$type=="five_prime_utr",] #subset to startcodons 
five_prime <- five_prime[five_prime$gene_id %in% GENES] # subset to all genes in DGE

#get enhancer in the surrounding locus
locus <- reduce(flank(five_prime[five_prime$gene_id == testgene],distance, both=T))
#locus <- flank(locus, distance,both=T) #reduce it , so no doubles appear
#locus<-reduce(locus)
locus_eCRMs<- g_eCRM[queryHits(findOverlaps(g_eCRM, locus, type = "any")),] #get CRMs overlapping that window

#matrices for glmnet
mat_CRM <- eDCA[locus_eCRMs$name,] #matrix of enhancers for glmnet
mat_GENE <- DGE[testgene,] #matrix GEX gene for glmnet

cvfit<- cv.glmnet(x=t(mat_CRM),y=t(log(mat_GENE))) #fit the model

# rank the CRMs according to the model
coef_CRM <- coef(cvfit, s = "lambda.min")
coef_CRM<-as.data.frame(coef_CRM[-1,]) 
colnames(coef_CRM) <- "coef"
coef_CRM <- coef_CRM[order(coef_CRM$coef,decreasing = T),,drop=FALSE]
#vi(cvfit)
coef_CRM
```

#get the putative promoter for the target gene
```{r,echo=F}
#promoters
pCRM_prom <- g_pCRM$name[g_pCRM$closGene1 ==testgene]
```


```{r,echo=F}
pl_embryos(patterns=DGE[testgene,],coordinates =  locations,title=paste("GEX of",translateG(testgene,TF_list)),type="RNA",psize=2)

pl_embryos(patterns=pDCA[pCRM_prom,],locations, title=paste("promoter of",translateG(testgene,TF_list)),type="ATAC",psize = 2)
#ggsave(filename = file.path("knirps_promoters.pdf"), device='pdf', dpi=400, useDingbats=FALSE)

pl_embryos(patterns=eDCA[rownames(coef_CRM),],coordinates =  locations ,type = "ATAC",psize = 1,nrow=5)
ggsave(filename = file.path("mef2_enhancers_rank.pdf"), device='pdf', dpi=400, useDingbats=FALSE)
#pl_fig_embryos(patterns=eDCA[rownames(coef_CRM),],coordinates =  locations,title=paste("CRMS sorted by coefficiants:"),type="ATAC")

```


R version 3.6.0 (2019-04-26)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.6 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] vip_0.3.2            Biostrings_2.54.0    XVector_0.26.0       ggplot2_3.3.3        dplyr_1.0.5          scales_1.2.0         rtracklayer_1.46.0   GenomicRanges_1.38.0
 [9] GenomeInfoDb_1.22.1  IRanges_2.20.2       S4Vectors_0.24.4     BiocGenerics_0.32.0  glmnet_4.1-1         Matrix_1.3-2        

loaded via a namespace (and not attached):
 [1] pillar_1.8.0                compiler_3.6.0              bitops_1.0-6                iterators_1.0.13            tools_3.6.0                 zlibbioc_1.32.0            
 [7] gtable_0.3.0                tibble_3.1.8                lifecycle_1.0.0             lattice_0.20-41             pkgconfig_2.0.3             rlang_1.0.3                
[13] foreach_1.5.1               DBI_1.1.1                   DelayedArray_0.12.3         cli_3.3.0                   rstudioapi_0.13             xfun_0.30                  
[19] gridExtra_2.3               GenomeInfoDbData_1.2.2      withr_2.4.1                 knitr_1.38                  generics_0.1.3              vctrs_0.4.1                
[25] tidyselect_1.1.2            grid_3.6.0                  glue_1.6.2                  Biobase_2.46.0              R6_2.5.0                    fansi_0.4.2                
[31] XML_3.99-0.3                survival_3.2-10             BiocParallel_1.20.1         sessioninfo_1.1.1           purrr_0.3.4                 magrittr_2.0.1             
[37] Rsamtools_2.2.3             codetools_0.2-16            matrixStats_0.58.0          splines_3.6.0               GenomicAlignments_1.22.1    assertthat_0.2.1           
[43] SummarizedExperiment_1.16.1 shape_1.4.5                 colorspace_2.0-0            utf8_1.2.1                  RCurl_1.98-1.3              munsell_0.5.0              

